### PR TITLE
Attention mask handling improvements

### PR DESCRIPTION
**Pull request by danieldk**
- Add the `AttentionMask` class, which stores a boolean mask and creates a logit mask on-demand. The logit mask is cached to avoid recomputing it for every layer.
- Make the attention mask mandatory in internal classes.
- Move attention mask check and view logic to
`ScaledDotProductAttention.forward`.
- Fix inconsistent naming as `mask`.

### CODE REVIEW

File: curated_transformers/models/attention.py [hunk 1]
```diff
@@ -1,10 +1,37 @@
...
+            raise ValueError(
+                f"attention mask of dtype torch.bool expected, was {self.bool_mask.dtype}"
+            )
+
+    @property
+    def logit_mask(self) -> Tensor:
+        if self._logit_mask is None:
+            # The value is `torch.finfo(attn_scores.dype).min`. Unfortunately,
+            # we cannot use `torch.finfo` in TorchScript.
+            self._logit_mask = (1.0 - self.bool_mask.int()) * -3.4028234663852886e38
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:25]: Can't we get the minimum value from the built-in `float` type (instead of using the constant directly)?

Comment (danieldk) [reply curated_transformers/models/attention.py:25]: `float` is double precision and we need single precision.

**Review by shadeMe — APPROVED** [review]
